### PR TITLE
Room dir: New filtering & 3rd party networks

### PR DESCRIPTION
**Pull request by dbkr**
Changes filtering on 3rd party networks to divide into portal / non portal rooms and not show portal rooms by default. Adds a special '_matrix' network for all rooms that aren't portal rooms.

Also adds ability to query 3rd party directory servers.

Fixes https://github.com/vector-im/vector-web/issues/2218
Requires https://github.com/matrix-org/matrix-react-sdk/pull/501

### CODE REVIEW

**Comment by dbkr** [comment]
Tests failing because of react PR requirement

File: src/components/structures/RoomDirectory.js [hunk 1]
```diff
@@ -365,14 +371,30 @@ module.exports = React.createClass({
...
-    _isRoomInNetwork(room, network) {
-        if (room.aliases && this.networkPatterns[network]) {
-            for (const alias of room.aliases) {
-                if (this.networkPatterns[network].test(alias)) return true;
+    _isRoomInNetwork(room, server, network) {
+        // We carve rooms into two categories here. 'portal' rooms are
+        // rooms created by a user joining a bridge 'portal' alias to
+        // participate in that room or a foreign network. A room is a
+        // portal room if it has exactly one alias and that alias matches
+        // a pattern defined in the config. It's network is the key
```
Comment (richvdh) [inline src/components/structures/RoomDirectory.js]: its

File: src/components/structures/RoomDirectory.js [hunk 2]
```diff
@@ -365,14 +371,30 @@ module.exports = React.createClass({
...
+        // of the pattern that it matches.
+        // All other rooms are considered 'native matrix' rooms, and
+        // go into the special '_matrix' network.
+
+        let roomNetwork = '_matrix';
+        if (room.aliases && room.aliases.length == 1) {
+            if (this.props.config.serverConfig && this.props.config.serverConfig[server] && this.props.config.serverConfig[server].networks) {
+                for (const n of this.props.config.serverConfig[server].networks) {
+                    const pat = this.networkPatterns[n];
+                    if (pat && pat) {
```
Comment (richvdh) [inline src/components/structures/RoomDirectory.js]: http://thedailywtf.com/articles/The-Test-of-Truth

File: src/components/structures/RoomDirectory.js [hunk 3]
```diff
@@ -365,14 +371,30 @@ module.exports = React.createClass({
...
+                        if (this.networkPatterns[n].test(room.aliases[0])) {
```
Comment (richvdh) [inline src/components/structures/RoomDirectory.js]: s/this.networkPatterns[n]/pat/

File: vector/config.sample.json [hunk 4]
```diff
@@ -6,30 +6,29 @@
...
-        "networks": [
-            "matrix:example_com",
-            "matrix:matrix_org",
-            "gitter",
-            "irc:freenode",
-            "irc:mozilla"
+        "servers": [
+            "matrix.org"
         ],
+	"serverConfig": {
```
Comment (richvdh) [inline vector/config.sample.json]: s/tab/space/

File: src/components/views/directory/NetworkDropdown.js [hunk 5]
```diff
@@ -26,19 +27,27 @@ export default class NetworkDropdown extends React.Component {
...
         };
     }
 
     componentWillMount() {
         // Listen for all clicks on the document so we can close the
         // menu when the user clicks somewhere else
         document.addEventListener('click', this.onDocumentClick, false);
+
+        // fire this now so the defaults can be set up
+        this.props.onOptionChange(this.state.selectedServer, this.state.selectedNetwork);
```
Comment (richvdh) [inline src/components/views/directory/NetworkDropdown.js]: so... RoomDirectory renders, which mounts a NetworkDropdown, which changes the state of RoomDirectory, which makes RoomDirectory render...

I'm slightly surprised react doesn't bitch about it, but I guess if it works, it's ok.

File: src/components/views/directory/NetworkDropdown.js [hunk 6]
```diff
@@ -90,41 +111,86 @@ export default class NetworkDropdown extends React.Component {
...
+            key += '_' + network;
+        }
+
+        return <div key={key} className="mx_NetworkDropdown_networkoption" onClick={click_handler}>
             {icon}
             <span className={span_class}>{name}</span>
         </div>;
     }
 
+    componentDidUpdate() {
```
Comment (richvdh) [inline src/components/views/directory/NetworkDropdown.js]: think our convention is to collect the react handlers at the top, before the private methods.

File: src/components/views/directory/NetworkDropdown.js [hunk 7]
```diff
@@ -90,41 +111,86 @@ export default class NetworkDropdown extends React.Component {
...
+        if (this.state.expanded && this.inputTextBox) {
+            this.inputTextBox.focus();
+        }
+    }
+
     render() {
-        const current_value = this._optionForNetwork(this.state.selectedNetwork, false);
+        let current_value = this._makeMenuOption(
```
Comment (richvdh) [inline src/components/views/directory/NetworkDropdown.js]: so this is thrown away if `this.state.expanded == true` ?

Put it in an `else` clause to save creating it?

**Comment by dbkr** [comment]
ptal

**Pull request merged**